File: api/js/src/validator/rules/inconsistent-configuration.ts
import { createDiagnostic } from "../diagnostics.js";
import type { ValidationRule } from "../types.js";
import { RULE_IDS } from "./constants.js";
import { loc } from "./helpers.js";

const ALLOWED_STYLES = new Set(["primary", "danger", "success"]);

export const inconsistentConfigurationRule: ValidationRule = {
  id: RULE_IDS.INCONSISTENT_CONFIGURATION,
  description: "Detects invalid builder or button configuration",
  defaultSeverity: "warning",
  run(ctx) {
    const diagnostics = [];

    if (!Number.isFinite(ctx.buttonsPerRow) || ctx.buttonsPerRow < 1) {
      diagnostics.push(
        createDiagnostic(
          RULE_IDS.INCONSISTENT_CONFIGURATION,
          `buttonsPerRow must be >= 1 (got ${ctx.buttonsPerRow})`,
          "error",
        ),
      );
    }

    if (ctx.autoWrapMaxChars < 0) {
      diagnostics.push(
        createDiagnostic(
          RULE_IDS.INCONSISTENT_CONFIGURATION,
          "autoWrapMaxChars cannot be negative",
          "error",
        ),
      );
    }

    for (const ref of ctx.normalized.flat) {
      const { button, rowIndex, columnIndex, flatIndex } = ref;
      if (
        button.style !== undefined &&
        !ALLOWED_STYLES.has(button.style)
      ) {
        diagnostics.push(
          createDiagnostic(
            RULE_IDS.INCONSISTENT_CONFIGURATION,
            `Invalid button style: ${String(button.style)}`,
            "error",
            loc(rowIndex, columnIndex, flatIndex, "style"),
            "Allowed: primary, danger, success",
          ),
        );
      }
    }

    return diagnostics;
  },
};

File: api/python/telegram_inline_keyboard_builder/validator/rules/inconsistent_configuration.py
from __future__ import annotations

import math

from ..diagnostics import create_diagnostic
from ..types import Diagnostic, RuleContext, ValidationRule
from .constants import RULE_IDS
from .helpers import loc

_ALLOWED_STYLES = frozenset({"primary", "danger", "success"})


def _run(ctx: RuleContext) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []

    if not math.isfinite(ctx.buttons_per_row) or ctx.buttons_per_row < 1:
        diagnostics.append(
            create_diagnostic(
                RULE_IDS.INCONSISTENT_CONFIGURATION,
                f"buttonsPerRow must be >= 1 (got {ctx.buttons_per_row})",
                "error",
            )
        )

    if ctx.auto_wrap_max_chars < 0:
        diagnostics.append(
            create_diagnostic(
                RULE_IDS.INCONSISTENT_CONFIGURATION,
                "autoWrapMaxChars cannot be negative",
                "error",
            )
        )

    for ref in ctx.normalized.flat:
        style = ref.button.get("style")
        if style is not None and style not in _ALLOWED_STYLES:
            diagnostics.append(
                create_diagnostic(
                    RULE_IDS.INCONSISTENT_CONFIGURATION,
                    f"Invalid button style: {style}",
                    "error",
                    loc(ref.row_index, ref.column_index, ref.flat_index, "style"),
                    "Allowed: primary, danger, success",
                )
            )

    return diagnostics


inconsistent_configuration_rule = ValidationRule(
    id=RULE_IDS.INCONSISTENT_CONFIGURATION,
    description="Detects invalid builder or button configuration",
    default_severity="warning",
    run=_run,
)
